Enables infinite precision on Money attributes based on column type

marcusmateus opened the pull request:
Uses work on `money` gem (https://github.com/ShippingEasy/money/pull/2) and the `monetize` gem (https://github.com/ShippingEasy/monetize/pull/1) to add infinite_precision support at the instance level.  Leveraging that work these changes set the precision on the Money instance for a model attribute based on that attributes backing column type.  If it is backed by a `:decimal` column an infinite precision Money instance is created.  Otherwise a normal precision instance is created.

I considered allowing the precision to be passed as an option on `monetize`, but it seemed awkward to have the precision not match the backing column type.  If you suddenly want more / less precision, you can simply change the column.  Perhaps that was overly prescriptive though.  Similarly, the current approach does not respect the global setting for precision `Money.default_infinite_precision` and instead favors the precision implied by the backing column type.

marcusmateus committed:
Enables infinite precision on Money attributes based on column type

Uses work on `money` gem (https://github.com/ShippingEasy/money/pull/2)
and the `monetize` gem (https://github.com/ShippingEasy/monetize/pull/1)
to add infinite_precision support at the instance level.  Leveraging
that work these changes set the precision on the Money instance for a
model attribute based on that attributes backing column type.  If it is
backed by a :decimal column an infinite precision Money instance is
created.  Otherwise a normal precision instance is created.

joshuaflanagan commented on lib/money-rails/active_record/monetizable.rb:
```diff
@@ -55,6 +55,9 @@ def monetize(*fields)
             end
 
+            # Assigns infinite_precision based on the field's column type, if available
+            options[:infinite_precision] = self.columns_hash[subunit_name].try(:type) == :decimal
```
Why store this in the `options` hash, just to slice it out later? Is there any code that currently accepts the `options` hash, that needs the `:infinite_precision` value? Seems like `options` is used to pass values _into_ this method.

How about:

```suggestion
            column_definition = columns_hash[subunit_name]
            money_options = {
              :infinite_precision => !!(column_definition && column_definition.type == :decimal)
            }
```

and then pass `money_options` to Money constructors.

joshuaflanagan replied:
Updated suggestion to remove use of ActiveSupport's `try`, since its behavior is confusing (and not usually what the author intended - see `try!`).

marcusmateus replied:
The options hash was reused for two reasons:
1. I initially had an option to set an attribute to `infinite_precision` directly that took precedence over the column type.  As you pointed out in our Slack convo the backing column could be a string, in which case we might want to allow setting the option directly after all.
2. To avoid changing the arity of the `read_monetized` and `write_monetized` methods... however, they do appear to be effectively private to the class, so changing the arity may not be an issue.

joshuaflanagan replied:
1. Yes, I think you probably should allow overriding, and only infer the value from `columns_hash` if it hasn't been specified.

2. I see now. I didn't realize use of the `options` hash spanned across multiple methods.

marcusmateus replied:
Implemented per our Slack convo with backwards compatibility and an `infer_precision` boolean option.

joshuaflanagan commented on spec/active_record/monetizable_spec.rb:
```diff
@@ -12,7 +12,9 @@ class Sub < Product; end
                      reduced_price_cents: 1500, reduced_price_currency: :lvl,
-                     lambda_price_cents: 4000)
+                     lambda_price_cents: 4000,
+                     unit_cost_cents: 1234.8765
```
It would be good to pass a fractional value to an existing column that is not stored as a decimal (like `price_cents`), to demonstrate that it will still round values as it used to.

```
price_cents: 3000.14
```

marcusmateus replied:
Actually, `price_cents` is the direct accessor for the column, so it will truncate.  But, using the accessor created by `monetize` should round, so I will add `price: 30.0051` which should get rounded to `3001` in cents.

marcusmateus replied:
Incorporated above into the new precision setting tests.

marcusmateus committed:
Adds explicit option to `monetize` for setting precision

Also provides a setting `infer_precision` that will tell `monetize` to
base the precision on the backing column type if it is not explicitly
provided.

For backwards compatability the feature works as follows:
  1. If an explicit option is provided to `monetize` that is used
  2. If no explicit option, and `infer_precision` is not set (defaults
     to false), then use the global `default_infinite_precision`.
  3. If no explicit option, and `infer_precision` *is* set, use
     `infinite_precision` if the attribute's column type is `:decimal`.

joshuaflanagan commented on lib/money-rails/active_record/monetizable.rb:
```diff
@@ -55,6 +55,16 @@ def monetize(*fields)
+                column_definition = self.columns_hash[subunit_name]
+                options[:infinite_precision] = column_definition && column_definition.type == :decimal
+              else
+                options[:infinite_precision] = MoneyRails.default_infinite_precision
```
Is there any difference in behavior between this line, and not sending a value for the option at all?
If no option is specified, wont `Money` default to `Money.default_infinite_precision`?

marcusmateus replied:
Yeah, you're right.  It is unnecessary since `MoneyRails.default_infinite_precision` delegates to `Money.default_infinite_precision`.  If it were different then it would be needed, but it is not.

marcusmateus replied:
Removed unnecessary assignment.

marcusmateus committed:
Removes unnecessary options[:infinite_precision] assignment

If `options[:infinite_precision]` is not set the default will
automatically be used.  There is no need to assign it explicitely.

marcusmateus committed:
Only introspect column type when backing table exists

Precision inference only works if the model's backing table already
exists.  If it does not, we are likely in a migration.
